src/MediatR/Mediator.cs

```diff
@@ -31,193 +30,122 @@ public Mediator(SingleInstanceFactory singleInstanceFactory, MultiInstanceFactor
         {
             var notificationHandlers = GetNotificationHandlers(notification)
-                .Select(handler => handler.Handle(notification, cancellationToken))
-                .ToArray();
-
-            return Task.WhenAll(notificationHandlers);
-        }
-
-        private RequestHandlerWrapper GetHandler(IRequest request)
-        {
-            return GetVoidHandler<RequestHandlerWrapper>(request,
-                typeof(IRequestHandler<>),
-                typeof(RequestHandlerWrapperImpl<>));
-        }
-
-        private RequestHandlerWrapper<TResponse> GetHandler<TResponse>(IRequest<TResponse> request)
-        {
-            return GetHandler<RequestHandlerWrapper<TResponse>, TResponse>(request,
-                typeof(IRequestHandler<,>),
-                typeof(RequestHandlerWrapperImpl<,>));
-        }
-
-        private AsyncRequestHandlerWrapper GetHandler(IAsyncRequest request)
-        {
-            return GetVoidHandler<AsyncRequestHandlerWrapper>(request,
-                typeof(IAsyncRequestHandler<>),
-                typeof(AsyncRequestHandlerWrapperImpl<>));
-        }
-
-        private AsyncRequestHandlerWrapper<TResponse> GetHandler<TResponse>(IAsyncRequest<TResponse> request)
-        {
-            return GetHandler<AsyncRequestHandlerWrapper<TResponse>, TResponse>(request,
-                typeof(IAsyncRequestHandler<,>),
-                typeof(AsyncRequestHandlerWrapperImpl<,>));
-        }
-
-        private CancellableAsyncRequestHandlerWrapper GetHandler(ICancellableAsyncRequest request)
-        {
-            return GetVoidHandler<CancellableAsyncRequestHandlerWrapper>(request,
-                typeof(ICancellableAsyncRequestHandler<>),
-                typeof(CancellableAsyncRequestHandlerWrapperImpl<>));
-        }
-
-        private CancellableAsyncRequestHandlerWrapper<TResponse> GetHandler<TResponse>(ICancellableAsyncRequest<TResponse> request)
-        {
-            return GetHandler<CancellableAsyncRequestHandlerWrapper<TResponse>, TResponse>(request,
-                typeof(ICancellableAsyncRequestHandler<,>),
-                typeof(CancellableAsyncRequestHandlerWrapperImpl<,>));
-        }
-
-        private TWrapper GetHandler<TWrapper, TResponse>(object request, Type handlerType, Type wrapperType)
-        {
-            var requestType = request.GetType();
-
-            var genericHandlerType = _genericHandlerCache.GetOrAdd(handlerType, new ConcurrentDictionary<Type, Type>())
-                .GetOrAdd(requestType, handlerType, (type, root) => root.MakeGenericType(type, typeof(TResponse)));
-            var genericWrapperType = _wrapperHandlerCache.GetOrAdd(wrapperType, new ConcurrentDictionary<Type, Type>())
-                .GetOrAdd(requestType, wrapperType, (type, root) => root.MakeGenericType(type, typeof(TResponse)));
+                .Select(handler => Task.Run(() => handler.Handle(notification), cancellationToken));

```

Comment:
Was this `Task.Run` intentional? I think we should steer as far away from `Task.Run` as possible.